url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPLinkManager.java

```diff
@@ -0,0 +1,2057 @@
+                || tempPort > CONFIG_LOCAL_PORT_MAX) {
+            throw new IllegalArgumentException("The local port is not valid.");
+        }
+        if (tempPort != localPort) {
+            // Local port has changed. Restart required.
+            localPort = tempPort;
+            significantChange = true;
+        }
+
+        String localFQDN;
+        try {
+            localFQDN = InetAddress.getLocalHost().getCanonicalHostName();
+        } catch (UnknownHostException e) {
+            // Cannot resolve the local host
+            localFQDN = "";
+        }
+        unitName = prop.getProperty(CONFIG_UNIT_NAME, localFQDN);
+        if (unitName.length() > CONFIG_UNIT_NAME_MAX) {
+            unitName = unitName.substring(0, CONFIG_UNIT_NAME_MAX);
+        }
+
+        log.setLevel(Level.toLevel(prop.getProperty("ocpSystemLoggingLevel",
+                "WARN")));
+
+        return significantChange;
+    }
+
+    /**
+     * Reloads the configuration, restarting the link if necessary.
+     *
+     * @param properties
+     *            The properties object to use
+     * @return true, if the configuration has changed enough to require a link
+     *         restart to fully apply (note that the link will not have been
+     *         restarted by this method)
+     */
+    public final boolean reloadConfig(final Properties properties) {
+        log.warn("reloadConfig Link " + linkIndex + " From " + properties);
+        if (loadConfig(properties)) {
+            log.warn("Link " + linkIndex + ": Restart required");
+            return true;
+        }
+        return false;
+    }
+    /**
+     * Clean up the OCP Link Manager. This method disconnects from the remote
+     * server, removes all registered handlers and de-registers all statistics.
+     * After calling this method the object can no longer be used.
+     */
+    protected final void cleanup() {
+        if (cleanedUp) {
+            return;
+        }
+
+        disconnect();
+
+        stats.unregisterStat(statPrefix + STAT_BAD_MESSAGE);
+        stats.unregisterStat(statPrefix + STAT_OUTGOING_HEARTBEAT);
+        stats.unregisterStat(statPrefix + STAT_CONNECTION_SUCCESS);
+        stats.unregisterStat(statPrefix + STAT_CONNECTION_FAIL);
+        stats.unregisterStat(statPrefix + STAT_MASTER_SLAVE_SWAP);
+
+        EnumSet<LegacyOCPMessageTypes> typeSet =
+                EnumSet.allOf(LegacyOCPMessageTypes.class);
+
+        for (LegacyOCPMessageTypes type : typeSet) {
+            if (type.getSupportRx()) {
+                stats.unregisterStat(statPrefix + STAT_MESSAGE_RX_BASE
+                        + type.getImplementation().getSimpleName());
+            }
+            if (type.getSupportTx()) {
+                stats.unregisterStat(statPrefix + STAT_MESSAGE_TX_BASE
+                        + type.getImplementation().getSimpleName());
+            }
+        }
+        tidHandlers.clear();
+        linkStateHandlers.clear();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **610**. At this line the following field is not protected by synchronization: `linkStateHandlers`. Your code typically protects this field with synchronization on object: `linkStateHandlers`.

**Fix**  
 Consider protecting the above field with synchronization on object: `linkStateHandlers`. For example, like at lines: **954**, **941**, **963**, **819**, **845**, **878**, and **863**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **819**, **845**, **863**, **878**, **941**, **954**, and **963**.